File: storage/db-core/src/main/java/com/coffee/gu/payment/PaymentJpaRepository.java
package com.coffee.gu.payment;

import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;

import java.util.Optional;

public interface PaymentJpaRepository extends JpaRepository<PaymentEntity, Long> {
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM PaymentEntity p WHERE p.id = :paymentId")
    Optional<PaymentEntity> findByIdForUpdate(Long paymentId);
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM PaymentEntity p WHERE p.orderKey = :orderKey")
    Optional<PaymentEntity> findByOrderIdForUpdate(String orderKey);
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/payment/PaymentJpaRepository.kt
package com.coffee.gu.payment

import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.Optional

interface PaymentJpaRepository : JpaRepository<PaymentEntity, Long> {
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM PaymentEntity p WHERE p.id = :paymentId")
    fun findByIdForUpdate(@Param("paymentId") paymentId: Long): PaymentEntity?

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM PaymentEntity p WHERE p.orderKey = :orderKey")
    fun findByOrderIdForUpdate(@Param("orderKey") orderKey: String): PaymentEntity?
}
